### PR TITLE
F363 - Animate Queue Creation and Deletion (and map optimization)

### PR DESCRIPTION
**Pull request by creiser2**
Animate Queue Creation and Deletion

In order to do this map code has changed in the following ways:
          1. Locations[:] now a dictionary that is keyed by queueId, covers all Circles[], Markers[] and Queues[] arrays. 
          2. Map location listeners only unsubscribe when zooming out of state or city.
          3. Map no longer has optimistic render on queue creation.

### CODE REVIEW

File: CranberryQueue/Map/MapController.swift [hunk 1]
```diff
@@ -83,103 +81,168 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
+        circle.strokeColor = UIColor(red: 180/255, green: 180/255, blue: 180/255, alpha: 0.5)
+        circle.strokeWidth = 2
+        circle.zIndex = 0
+        circle.map = map
+        locations[location.queueId]!["marker"] = marker
+        locations[location.queueId]!["circle"] = circle
+    }
+    
+    // this function takes in an array of document changes and handles them accordingly
+    func processDocumentChanges(documentChanges: [DocumentChange]) {
```
Comment (JohnnyPersichilli) [inline CranberryQueue/Map/MapController.swift]: maybe add something in these function names to indicate changes in location docs incase we add similar functions for other docs in the future

File: CranberryQueue/Map/MapController.swift [hunk 2]
```diff
@@ -222,15 +285,9 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
-        
         curCoords = center
-        
         setupMap(withCoords: center)
-        getGeoCode(withLocation: ["lat": center.latitude, "long": center.longitude]){ city, region in
-            self.mapControllerDelegate?.updateGeoCode(city: city, region: region)
-            self.watchLocationQueues(city: city, region: region)
-        }
-        
+        mapView(map!, idleAt: GMSCameraPosition(latitude: center.latitude, longitude: center.longitude, zoom: 14))
```
Comment (JohnnyPersichilli) [inline CranberryQueue/Map/MapController.swift]: pretty sure we previously had the default zoom at 15 and now its 14 but thats not a big deal

File: functions/index.js [hunk 3]
```diff
@@ -270,3 +270,24 @@ exports.updateDownvotes = functions.firestore
     })
 
     });
+
```
Comment (JohnnyPersichilli) [inline functions/index.js]: take out excess whitespace in this section

File: CranberryQueue/Map/MapController.swift [hunk 4]
```diff
@@ -83,103 +81,168 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
+            if(self.curZoom > 10)   {
+                if self.city != city {
+                    self.removeAllLocations()
+                    self.getTopQueuesInCity(city: city, region: region)
+                }
+            } else {
+                if self.region != region {
+                    self.removeAllLocations()
+                    self.getTopQueuesInState(region: region, zoom: self.curZoom)
+                }
```
Comment (rolflocher) [inline CranberryQueue/Map/MapController.swift]: would this not query by state if you zoomed out on bryn mawr and didnt change your region?

Comment (rolflocher) [reply CranberryQueue/Map/MapController.swift]: hard to notice in practice bc all our queues are in one town

File: CranberryQueue/Map/MapController.swift [hunk 5]
```diff
@@ -83,103 +81,168 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
+    
+    // draw a marker and circle at a given location
+    func drawMarker(location: CQLocation) {
+        let circleCenter = CLLocationCoordinate2D(latitude: location.lat, longitude: location.long)
+        // class that allows creation animation
+        let circle = TCCircle(position: circleCenter, radius: 0.0)
+        let defaultColor = UIColor(red: 180/255, green: 180/255, blue: 180/255, alpha: 0.3)
+        let homeColor = UIColor(displayP3Red: 189/255, green: 209/255, blue: 199/255, alpha: 0.7)
+        circle.fillColor = location.queueId == self.queueId ? homeColor : defaultColor
+        let position = CLLocationCoordinate2D(latitude: location.lat, longitude: location.long)
```
Comment (rolflocher) [inline CranberryQueue/Map/MapController.swift]: this was already calculated above. circleCenter

File: CranberryQueue/Map/MapController.swift [hunk 6]
```diff
@@ -83,103 +81,168 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
+        let marker = GMSMarker(position: position)
+        circle.addCircleAnimation(to: 200.0, duration: 1, completion: {
+           //popup animation for markers
+           marker.appearAnimation = GMSMarkerAnimation.pop
+           marker.icon = location.queueId == self.queueId ? GMSMarker.markerImage(with: UIColor.green) : GMSMarker.markerImage(with: UIColor(displayP3Red: 145/255, green: 158/255, blue: 188/255, alpha: 1))
+           marker.title = location.name
+           marker.map = self.map
+           marker.userData = location
```
Comment (rolflocher) [inline CranberryQueue/Map/MapController.swift]: do you mind taking the appearAnimation out of the closure?

File: CranberryQueue/Map/MapController.swift (hunk 1 above)
Comment (rolflocher) [reply CranberryQueue/Map/MapController.swift]: or perhaps have the helper definition immediately preceding the last call to it. same with all these new helpers

File: CranberryQueue/Map/MapController.swift [hunk 7]
```diff
@@ -83,103 +81,168 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
-    func getTopQueusInCity(city: String, region: String) {
+    // this function iterates over the locations dictionary and removes all of them
+    func removeAllLocations() {
+        let keys = Array(locations.keys)
+        for key in keys {
+            let circle = locations[key]!["circle"] as! TCCircle
+            circle.fillColor = UIColor(red: 180/255, green: 0/255, blue: 0/255, alpha: 0.3)
+            let marker = locations[key]!["marker"] as! GMSMarker
+            DispatchQueue.main.async {
+                circle.removeCircleAnimation(from: 200, duration: 2, completion: {
```
Comment (rolflocher) [inline CranberryQueue/Map/MapController.swift]: can you put the fillcolor assignment on the main thread? i understand keeping it outside the completion

File: CranberryQueue/Map/MapController.swift [hunk 8]
```diff
@@ -321,3 +368,56 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
+          self.completion = completion
+          self.duration = duration
+          self.begin = NSDate()
+          self.performSelector(onMainThread: #selector(shrinkRadius), with: nil, waitUntilDone: false)
+    }
+
+    @objc func expandRadius() {
+        let i : TimeInterval = NSDate().timeIntervalSince(self.begin as Date)
+        
+        if (i >= self.duration) {
```
Comment (rolflocher) [inline CranberryQueue/Map/MapController.swift]: ws

File: CranberryQueue/Map/MapController.swift [hunk 9]
```diff
@@ -321,3 +368,56 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
+            self.radius = d
+            self.performSelector(onMainThread: #selector(expandRadius), with: nil, waitUntilDone: false)
+        }
+    }
+    
+    @objc func shrinkRadius() {
+        let i : TimeInterval = NSDate().timeIntervalSince(self.begin as Date)
+        if (i >= self.duration) {
+            self.radius = 0
+            completion?()
```
Comment (rolflocher) [inline CranberryQueue/Map/MapController.swift]: my favorite part

File: CranberryQueue/Map/MapController.swift (hunk 4 above)
Comment (creiser2) [reply CranberryQueue/Map/MapController.swift]: Yeah thats a bug. Thanks.

File: CranberryQueue/Map/MapController.swift [hunk 10]
```diff
@@ -83,103 +81,168 @@ class MapController: UIViewController, CLLocationManagerDelegate, GMSMapViewDele
...
+        if homeColor! {
+          marker.icon = GMSMarker.markerImage(with: UIColor.green)
         } else {
-            getTopQueusInState(region: region, zoom: curZoom)
+          marker.icon = GMSMarker.markerImage(with: UIColor(displayP3Red: 145/255, green: 158/255, blue: 188/255, alpha: 1))
+        }
+    }
+    
+    // when a locations data is modifed ex: numMembers or currentSong
+    func modifyLocation(diff: DocumentChange) {
```
[the remark this answers is not in the document]
Comment (creiser2) [reply CranberryQueue/Map/MapController.swift]: Sounds good.

File: CranberryQueue/Map/MapController.swift (hunk 1 above)
Comment (creiser2) [reply CranberryQueue/Map/MapController.swift]: @JohnnyPersichilli this function should work for any type of document change. It's not location specific.

**Pull request merged**